### ISSUE TITLE
HTTP/1.x transport connection broken

### ISSUE THREAD

**Issue by mkgvt** [minio/mc#3965]
## Expected behavior

The data is transferred without error.

## Actual behavior

Error message: `mc: <ERROR> Failed to copy  <file>. Put "https://<dst-ip>:9001/<file>": net/http: HTTP/1.x transport connection broken: http: ContentLength=341 with Body length 339`

## Steps to reproduce the behavior

The issue is not the network as was stated in #3075. It appears to be due to MinIO using the wrong file size for ContentLength. `mc` thinks the length of the file is 341 bytes but the length is only 339 bytes:

$ ls -l --block-size 1 <file>
0 -rw-r--r-- 1 user user 339 Feb  5 03:57 `<file>`

Note: this is the first time the issue has happened in 85 TiB transferred without error from <src> to <dst> since July 28 202.

Transfer command:

$ mc --insecure cp -q --md5 -a -r -c `<file-on-src>` `<dst-alias>/<dst-path>/`

## mc --version
- mc version RELEASE.2021-05-18T03-39-44Z

Note: a transfer of a 207 GiB tar archive of the directory in which the 339 byte file resides worked without error.

File statistics of files transferred to date (with only this error):
- Count: 290,138
- Minimum: 167 B
- Maximum: 222,051,665,920 B
- Mean: 318,816,670 B
- Standard Deviation: 810,712,195 B
- Median: 56,374,820

## System information

Sender:

$ uname -a
Linux <src> 2 3.10.0-693.21.1.el7.x86_64 #1 SMP Wed Mar 7 19:03:37 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

$ cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
...

Source file system is GPFS.

Receiver:

$ uname -a
Linux <dst> 5.4.0-90-generic #101-Ubuntu SMP Fri Oct 15 20:00:55 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

 cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.3 LTS (Focal Fossa)"
...

Destination file system is Ceph.

**Comment by harshavardhana** [comment]
Can you share `mc ls localfile` output?

**Comment by mkgvt** [comment]
`$ mc ls ​localfile
[2022-02-05 03:57:26 EST]   339B ​localfile`

**Comment by vadmeste** [comment]
> $ mc ls ​localfile [2022-02-05 03:57:26 EST] 339B ​localfile

I find it is hard to imagine this is an mc issue. Is the issue still reproducible for you?

**Comment by mkgvt** [comment]
Yes but I found a work around. I tar'd the whole directory, transferred the archive, then extracted the file.

**Comment by mkgvt** [comment]
How does `mc` determine the size of the file before it sends it? How does it get ContentLength? A hypothesis it that the file system misinforms `mc`. I'll see if I can verify if that is the case or not.

**Comment by vadmeste** [comment]
> How does mc determine the size of the file before it sends it? How does it get ContentLength? A hypothesis it that the file system misinforms mc. I'll see if I can verify if that is the case or not.

It is a simple file system stat. Is there any chance that the file got updated ? mc does listing first then uploads the file. The file could be changed meanwhile

**Comment by harshavardhana** [comment]
> It didn't change in between. I retransferred it a day later and it was still having the exact same error even though nothing changed. My current hypothesis (under the assumption that mc is exonerated) is the GPFS, which is the file system that localfile is on, is returning a different value from stat than it is when reading the file. I've had weird behavior from GPFS before. All parallel file systems play loose with POSIX semantics for performance but I've had more problems with GPFS than the others.

Can you do `mc cat <filename> | wc -c`

**Comment by mkgvt** [comment]
Note: I can't check that exact file anymore because I had to clean up space on that file system. So I have no way of testing the hypothesis. Closing. I'll open it again if it happens again so I have a means of continuing to investigate.
